ledger/forest/forest.go

```diff
@@ -0,0 +1,194 @@
+//  for our case and keep another one with more features around.
+
+// Forest is a forest of tries with an unlimited size, but which can be reset manually to remove all tries besides one.
+type Forest struct {
+	values map[ledger.Path]*ledger.Payload
+	steps  map[flow.StateCommitment]step
+}
+
+func New() *Forest {
+	f := Forest{
+		steps:  make(map[flow.StateCommitment]step),
+		values: make(map[ledger.Path]*ledger.Payload),
+	}
+
+	return &f
+}
+
+func (f *Forest) Add(tree *trie.Trie, paths []ledger.Path, payloads []*ledger.Payload, parent flow.StateCommitment) {
+	commit := flow.StateCommitment(tree.RootHash())
+
+	// FIXME: Is this still necessary? IIRC it was a temporary solution because we didn't have access to payloads anymore. Since we can list the paths and have access to the DB, we might do that instead of storing values in there.
+	for i := range paths {
+		if payloads == nil {
+			f.values[paths[i]] = nil
+		} else {
+			f.values[paths[i]] = payloads[i]
+		}
+	}
+
+	s := step{
+		tree:   tree,
+		parent: parent,
+	}
+	f.steps[commit] = s
+}
+
+func (f *Forest) Has(commit flow.StateCommitment) bool {
+	_, ok := f.steps[commit]
+	return ok
+}
+
+func (f *Forest) Tree(commit flow.StateCommitment) (*trie.Trie, bool) {
+	s, ok := f.steps[commit]
+	if !ok {
+		return nil, false
+	}
+
+	return s.tree, true
+}
+
+func (f *Forest) Parent(commit flow.StateCommitment) (flow.StateCommitment, bool) {
+	st, ok := f.steps[commit]
+	if !ok {
+		return flow.DummyStateCommitment, false
+	}
+
+	return st.parent, true
+}
+
+func (f *Forest) Values() map[ledger.Path]*ledger.Payload {
+	return f.values
+}
+
+func (f *Forest) Reset(finalized flow.StateCommitment) {
+	for commit := range f.steps {
+		if commit != finalized {
+			delete(f.steps, commit)
+		}
+	}
+}
+
+// FIXME: The following is only needed for this struct to be used as a flow forest.
+
+func (f *Forest) Read(r *ledger.TrieRead) ([]*ledger.Payload, error) {
+
+	if len(r.Paths) == 0 {
+		return []*ledger.Payload{}, nil
+	}
+
+	commit, err := flow.ToStateCommitment(r.RootHash[:])
+	if err != nil {
+		return nil, fmt.Errorf("invalid trie root hash: %w", err)
+	}
+
+	trie, ok := f.Tree(commit)
+	if !ok {
+		return nil, fmt.Errorf("trie does not exist (root hash: %x)", commit[:])
+	}
+
+	// Deduplicate keys.
+	deduplicatedPaths := make([]ledger.Path, 0, len(r.Paths))
+	pathOrgIndex := make(map[ledger.Path][]int)
+	for i, path := range r.Paths {
+		// only collect duplicated keys once
+		indices, ok := pathOrgIndex[path]
+		if !ok { // deduplication here is optional
+			deduplicatedPaths = append(deduplicatedPaths, path)
+		}
+		// append the index
+		pathOrgIndex[path] = append(indices, i)
+	}
+
+	payloads := trie.UnsafeRead(r.Paths)
+
+	// Reconstruct the payloads in the same key order as the given paths.
+	orderedPayloads := make([]*ledger.Payload, len(r.Paths))
+	totalPayloadSize := 0
+	for i, p := range deduplicatedPaths {
+		payload := payloads[i]
```

Comment:
No, since the payloads are fetched from the deduplicated list, we'll get `payloadA, payloadB, payloadC` and so on. The duplicate has been removed before the call.

TBH I am also not fond of how this is implemented, I just kept it the way they did it, but I will rewrite it taking your considerations in mind, if we even need to have the `Read` method in the end 👍

Reply:
@Maelkum Does this solve it for you or did I miss something?

Reply:
I think in the original code the issue was that it was the duplicated list that was used for retrieval of payloads, that's why I thought there'd be an issue - we'd pairing deduplicated paths with duplicated payloads by index.

But in any case, that's gone now and so is the conundrum :D